text/0156-columnar-encoding.md

```diff
@@ -0,0 +1,1417 @@
+This `batch_id` will be used in the `BatchStatus` object to acknowledge receipt and processing of the corresponding
+batch.
+A simple numeric counter can be used to implement this batch_id, the goal being to use the most concise id possible.
+
+### Substream ID Generation
+
+The `sub_stream_id` attribute is used to identify a sub-stream of `BatchArrowRecords` that share the same characteristics.
+The life cycle of a substream is as follows:
+
+* In addition to the data, the first `BatchArrowRecords` will contain its schema and an optional set of dictionaries.
+  A `sub_stream_id` is created and associated to this `BatchArrowRecords`.
+* The following `BatchArrowRecords` sharing the same characteristics then refer to the `sub_stream_id` to avoid the
+  retransmission of the schema and the dictionaries. Concerning the dictionaries, it is however possible to transmit an
+  updated version with this mechanism (i.e. delta dictionaries).
+
+Multiple approaches are possible to create this `sub_stream_id` depending on the producer and the way the telemetry data
+is generated. Depending on the context, certain approaches are more appropriate:
+
+* When the producer is already organized to produce uniform streams then a simple numeric counter can be used.
+* A more generic (but less efficient) approach is to generate a hash from the columns names and types (sorted
+  lexicographically)
+  and use this hash as `sub_stream_id`. This allows all schema-compatible telemetry data to be sent on the same
+  substream. Although collisions at this scale will be extremely rare, it is advisable to implement a strategy to avoid
+  hash collisions (e.g. keep a map of hash -> [(suffix, schema_id), ...])
+
+### Schema ID Generation
+
+Within the collector, batching, filtering, exporting, ... operations require to group the `BatchArrowRecords` having a
+compatible schema. A synthetic identifier (or `schema_id`) must be computed for each `BatchArrowRecords` to perform this
+grouping.
+
+Since batch events come from multiple and uncontrolled sources, it is not possible to generalize the use of
+the `sub_stream_id`
+outside the connection between the source and the collector. There is no guarantee that all sources encode the
+`sub_stream_id` in the same way. Therefore, we recommend calculating the schema id in the following way:
+
+* for each Arrow Schema create a list of triples (name, type, metadata) for each column.
+* sort these triples according to a lexicographic order.
+* concatenate the sorted triples with a separator and use these identifiers as `schema_id` (or a shorter version via
+  an equivalence table).
+
+### Traffic Balancing Optimization
+
+To mitigate the usual pitfalls of a stream-oriented protocol, protocol implementers are advised to:
+
+* client side: create several streams in parallel (e.g. create a new stream every 10 event types),
+* server side: close streams that have been open for a long time (e.g. close stream every 1 hour).
```

Comment:
Server closing the stream results in the client not knowing what happened to unacknowledged BatchArrowRecords. I think client closing the stream is the safer approach.

Reply:
I've tried to explain above how that the arrow mechanism performs an immediate retry internally for this condition when a stream is closed due to connection aging. The caller will wait while the request is tried on a new stream, which will start as soon as an aging stream exits.

Also note that the Collector's gRPC configuration already includes the keepalive parameters including `MaxAge` and `MaxAgeGrace` so I believe all we need to do is document that these are good settings to use with streaming. We could force these settings, and I suspect we will see requests to limit streams based on memory consumed as well as time, so this is an area for refinement.